chatApp/chatApp/clientServer.cpp

```diff
@@ -0,0 +1,282 @@
+			}
+			else
+			{
+				//char buf[4096];
+				ZeroMemory(buf, MAX_BUFFER_SIZE);
+
+				//Receive message 
+				int bytesIn = recv(sock, buf, MAX_BUFFER_SIZE, 0);
+				if (bytesIn <= 0)
+				{
+					//drop the client 
+					closesocket(sock);
+					FD_CLR(sock, &master);
+				}
+				else
+				{
+					// Send message to other clients, and definiaetly NOT listening socket
+					for (int i = 0; i < master.fd_count; i++)
+					{
+						SOCKET outSock = master.fd_array[i];
+						if (outSock != listening && outSock != sock)
+						{
+							std::ostringstream ss;
+							ss << "SOCKET #" << sock << ": " << buf << "\r";
+							std::string strOut = ss.str();
+							send(outSock, strOut.c_str(), strOut.size() + 1, 0);
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+void TCPserver::sendMsg(int clientSocket, std::string msg)
+{
+	send(clientSocket, msg.c_str(), msg.size() + 1, 0);
+}
+
+void TCPserver::cleanup()
+{
+	WSACleanup();
+}
+
+TCPclient::TCPclient(std::string ipAddress, int port)
+	: m_ipAddress(ipAddress), m_port(port)
+{
+}
+
+bool TCPclient::init()
+{
+	WSADATA wsData;
+
+	int wsResult;
+	wsResult = WSAStartup(MAKEWORD(2, 2), &wsData);
+
+	return wsResult == 0;
+}
+
+SOCKET TCPclient::createSocket()
+{
+	// Create a socket
+	SOCKET sock = socket(AF_INET, SOCK_STREAM, 0);
+	if (sock != INVALID_SOCKET)
+	{
+		sockaddr_in hint;
+		hint.sin_family = AF_INET;
+		hint.sin_port = htons(m_port);
+		inet_pton(AF_INET, m_ipAddress.c_str(), &hint.sin_addr);
+		return sock;
+	}
+	else
+	{
+		std::cerr << "Creating socket failed! - #" << WSAGetLastError() << std::endl;
+		closeSock(sock);
+		return -1;
+	}
+}
+
+int TCPclient::connectClientServer(SOCKET sock)
+{
+	sockaddr_in hint;
+	hint.sin_family = AF_INET;
+	hint.sin_port = htons(m_port);
+	inet_pton(AF_INET, m_ipAddress.c_str(), &hint.sin_addr);
+
+	int connResult = connect(sock, (sockaddr*)&hint, sizeof(hint));
+	if (connResult == SOCKET_ERROR)
+	{
+		std::cerr << "Connect to server failed! - Error #" << WSAGetLastError() << std::endl;
+		closeSock(sock);
+		return -1;
+	}
+	else
+		return connResult;
+}
+
+void TCPclient::run()
+{
+	char buf[MAX_BUFFER_SIZE];
+
+	//Create a listening socket
+	SOCKET sock = createSocket();
+		if (sock == INVALID_SOCKET)
+		{
+			closeSock(sock);
+		}
+		//Connect to server
+		connectClientServer(sock);
+	// Do-while loop to send and receive data 
+	std::string userInput;
+
+	do
+	{
+		// Promt the user for some text
+		std::cout << "> ";
+		getline(std::cin, userInput);
+		//std::cin >> userInput;
+
+		if (userInput.size() > 0)
+		{
+			// Send the text
+			int sendResult = send(sock, userInput.c_str(), userInput.size() + 1, 0);
+			if (sendResult != SOCKET_ERROR)
+			{
+				// Wait for response
+				ZeroMemory(buf, MAX_BUFFER_SIZE);
+				int bytesReceived = recv(sock, buf, MAX_BUFFER_SIZE, 0);
+				if (bytesReceived > 0)
+				{
+					// Echo response to console
+					std::cout << "SERVER " << std::string(buf, 0, bytesReceived) << std::endl;

```

Comment:
Pytanie, czy nie można by było jakoś polimorficznie upakować tego couta, tak by nie wywyoływać go bezpośrednio, ale zrobić klasę w sytlu presenter, która będzie mieć metodę do wyświetlania i pobierania danych i wtedy przekazać ją do Twojej klasy tutaj, tak by kod był wolny od cout i cin, bo np, jak będziesz mieć QT to tak będą inne metody odpowiedzialne za wyświetlanie, a tak teraz jesteś trochę uziemiony :)

Reply:
class TCPpresnter
{
	friend class TCPclient;
private:



public:
	virtual void echoResponse(char * buffer, int sizeRecv);
	virtual void promtUser(std::string user);
	virtual void welcomeServer(SOCKET sock);
	virtual void socketNameSender(SOCKET sock, char * buffer, SOCKET sockOut);

};


class TcpCommandLinePresenter : public TcpPresenter {
//metody z cin i cout
}

Reply:
TcpPresneter *pr = new TcpCommandLinePresenter();
pr->welcomeServer();// w tym miejscu korzystamy z metody welcomeServlet z TcpCommandLinePresenter
pr = new TcpGuiPresenter();
pr->welcomeServer();// w tym miejscu korzystamy z metody welcomeServlet TcpGuiPresenter